gum: add page (#8314)
tldr-pages/tldr: pages/common/gum.md

# gum

> A tool for making glamorous shell scripts.
> More information: <https://github.com/charmbracelet/gum>.

> - Interactively pick an option to print to `STDOUT`:

Reviewer: ```suggestion
- Interactively pick a specific option to print to `STDOUT`:
```

Reviewer 2: Is STDOUT in caps necessary?

Reviewer: I like this but our style guide doesn't provide any info about how we have to write stdin, stdout, stderr.

Author: Looking at what we currently use, lowercase is more used.

> `gum choose "{{option_1}}" "{{option_2}}" "{{option_3}}"`

Reviewer: ```suggestion
`gum choose "{{option1}}" "{{option2}}" "{{option3}}"`
```

Author: There can also be as many options, can we include a ...

Author: This is incorrect syntax, 

the each option must be encased with quotation marks.

What do you think of

```suggestion
`gum choose "{{option_1}}" "{{option_2}}" "{{option_3}}" "{{...}}"`
```

Reviewer: > the each option must be encased with quotation marks.

But it is [valid](https://github.com/tldr-pages/tldr/blob/main/pages/linux/zip.md) for directories:

```
{{path/to/directory1 path/to/directory2 ...}}
```

🤔

Author: $ gum choose "one two three"
> one two three

This is not what we want. We want it to be
> One 
> Two
> Three

since the current output gives one "option"

Reviewer: We can put quotes inside placeholder.

Reviewer 3: Let me watch this .. I am the most confused person when it comes to `placeholder` syntax :)
👀

Reviewer 4: @EmilySeville7cfg we always put quotes outside of the placeholders. Because the quote itself is not part of the argument, it is always required. In all our English pages, we have almost 700 matches for `"{{...}}"` and only 7 for `{{"..."}}` (which we should fix).

Reviewer: Got it.

> - Open an interactive prompt for the user to input a string with a placeholder:

Reviewer: ```suggestion
- Open an interactive prompt for the user to input a string with a specific placeholder:
```

Reviewer 2: Hey, Shouldn't this line be
```suggestion
- Open an interactive prompt for the user to input a commit message:
``` 
[Reference](https://github.com/charmbracelet/gum#readme)
![image](https://user-images.githubusercontent.com/26346867/183278051-adcfe256-41d9-4525-ba7f-cff1ba11c4de.png) 
![image](https://user-images.githubusercontent.com/26346867/183278201-9a13595d-d1d2-4ce0-8d93-3d28550e448b.png)

Author: > Hey, Shouldn't this line be
> 
> [Reference](https://github.com/charmbracelet/gum#readme) ![image](https://user-images.githubusercontent.com/26346867/183278051-adcfe256-41d9-4525-ba7f-cff1ba11c4de.png) ![image](https://user-images.githubusercontent.com/26346867/183278201-9a13595d-d1d2-4ce0-8d93-3d28550e448b.png)

No, since the command doesn't need to be used for a commit message or whatever, that is just an example they provide.

> `gum input --placeholder "{{placeholder_string}}"`

Reviewer 2: ```suggestion
`gum input --placeholder "{{placeholder_value}}"`
```

> `gum input --placeholder "{{placeholder_string}}"`

Reviewer: ```suggestion
`gum input --placeholder "{{value}}"`
```

What about omit `placeholder` term from the placeholder? I don't like this repetition.

> - Interactvely pick an option which will exit if "No" is chosen:

Reviewer 4: Doesn't the command exit in both cases, just with different exit values?
```suggestion
- Open an interactive confirmation prompt and exit with either `0` or `1`:
```
I know, it is not limited to confirmation prompts, but that obviously what the subcommand was meant for

> `gum confirm "{{Continue?}}" --default=false --affirmative "{{Yes}}" --negative "{{No}}" {{&& : || exit 1}}`

Reviewer: ```suggestion
`gum confirm "{{Continue?}}" --default=false --affirmative "{{Yes}}" --negative "{{No}}" {{|| exit 1}}`
```

Author: IMO, yes since it gives the user an idea that you can add a command for if `Yes` is chosen.

Maybe we could change it to `{{&& echo "Yes selected" || echo "No selected"}}

Reviewer: > aybe we could change it to {{&& echo "Yes selected" || echo "No selected"}}

I like this variant.

- Show a spinner while a command is taking place with text alongside:

> `gum spin --spinner {{dot|line|minidot|jump|pulse|points|globe|moon|monkey|meter|hamburger}} --title "{{loading...}}" -- {{command}}`

Reviewer: I don't like too many options listed inside curly braces. I think we should not do that not to introduce extra complexity in tldr pages. Maybe we can limit alternatives up to 5 (inclusive).

Reviewer 2: Yeah, I too think the same. Maybe we can reduce it to some essential examples as displaying it would be complex for the client. 
(Or)
We can provide only one example as given in the [Github Page](https://github.com/charmbracelet/gum#spin).
```suggestion
`gum spin --spinner {{dot}} --title "{{title}}" -- {{command}}`
```

Author: see `tldr nmcli`, its better to just show all of the options since otherwise the user may not know that the option exists. The spinner choices aren't well documented from what I can see, I found them by going [here](https://github.com/charmbracelet/gum/blob/main/spin/spinners.go)

Reviewer: I think we can list all options just when docs don't describe them all and user have to dig into source code to understand what can be used.

> `gum spin --spinner {{dot|line|minidot|jump|pulse|points|globe|moon|monkey|meter|hamburger}} --title "{{loading...}}" -- {{command}}`

Reviewer: ```suggestion
`gum spin --spinner {{dot|line|minidot|jump|pulse|...}} --title "{{loading...}}" -- {{command}}`
```

Let's use `...` when there are too many alternatives to list.

Reviewer 2: Yeah, This seems better. It would fix the [issue](https://github.com/tldr-pages/tldr/pull/8314#discussion_r939615534).

Author: This could work, although as the user I have no idea of what the other options are and since they aren't obvious like when we use {[file1 file2 ...}}, I'd prefer to keep them.

Reviewer 3: In my opinion, our aim is to help the `user`; and if the options are not absolutely obvious and can fit on an example, we should have them.
Although, there is may be a balance that we folks might want to strike where we do not make it too verbose :)

Reviewer 4: This example with all possible options is exactly as long as the previous one. I think we can keep them all.

- Format text to include emojis:

> `echo "{{:smile: :heart hello}}" | gum format -t {{emoji}}`

Author: ```suggestion
`echo "{{:smile: :heart: hello}}" | gum format -t {{emoji}}`
```

> `echo "{{:smile: :heart: hello}}" | gum format -t {{emoji}}`

Reviewer 4: Why not like this? Not all subcommands accept taking the text from stdin.
```suggestion
`gum format -t {{emoji}} "{{:smile: :heart: hello}}"`
```

Author: sure, I see no issue
